packages/tokenbank_v1/contracts/impl/MetaTxModule.sol

```diff
@@ -0,0 +1,254 @@
+
+        address wallet = extractWalletAddress(data);
+        bytes32 metaTxHash = getSignHash(
+            wallet, // from
+            address(this),  // to
+            0, // value
+            data,
+            nonce,
+            gasPrice,
+            gasLimit,
+            gasToken,
+            "" // extraHash
+        );
+
+        require(
+            validateSignatures(wallet, data, metaTxHash, signatures),
+            "INVALID_SIGNATURES"
+        );
+
+        saveExecutedMetaTx(wallet, nonce, metaTxHash);
+        (bool success,) = address(this).call(data);
+
+        if (gasPrice > 0) {
+            uint gasSpent = startGas - gasleft();
+            require(gasSpent <= gasLimit, "EXCEED_GAS_LIMIT");
+
+            gasSpent = gasSpent.mul(gasPrice).add(GAS_OVERHEAD);
+            require(
+                Wallet(wallet).transferToken(msg.sender, gasSpent, gasToken),
+                "OUT_OF_GAS"
+            );
+        }
+
+        emit ExecutedMetaTx(wallet, nonce, metaTxHash, success);
+    }
+
+    /// @dev Returns the last nonce used by a wallet.
+    /// @param wallet The wallet's address.
+    /// @return Last nonce used.
+    function lastNonce(address wallet)
+        public
+        view
+        returns (uint)
+    {
+        return wallets[wallet].nonce;
+    }
+
+    function extractWalletAddress(bytes memory data)
+        internal
+        pure
+        returns (address wallet)
+    {
+        require(data.length >= 36, "INVALID_DATA");
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            // data layout: {length:32}{sig:4}{_wallet:32}{...}
+            wallet := mload(add(data, 36))
+        }
+    }
+
+    /// @dev calculate the hash that all signatures will sign against.
+    ///      See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1077.md
+    function getSignHash(
+        address from,
+        address to,
+        uint256 value,
+        bytes   memory data,
+        uint256 nonce,
+        uint256 gasPrice,
+        uint256 gasLimit,
+        address gasToken,
+        bytes   memory extraHash
+        )
+        internal
+        pure
+        returns (bytes32)
+    {
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                byte(0x19),
+                byte(0),
+                from,
+                to,
+                value,
+                keccak256(data),
+                nonce,
+                gasPrice,
+                gasLimit,
+                gasToken,
+                extraHash
+                )
+            );
+
+        return keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash));
+    }
+
+    /// @dev   Recovers the signer at a given index from a list of concatenated signatures.
+    /// @param metaTxHash The signed hash
+    /// @param signatures The concatenated signatures.
+    /// @param index The index of the signature to recover.
+    function recoverSigner(
```

Comment:
Should use the library function in `SignatureUtil`?

Reply:
Done.

Reply:
done